### ISSUE TITLE
Thread safety / unsafe code blocks

### ISSUE THREAD

**Issue by tomtau** [ChosunOne/merkle_bit#20]
# Development Process Improvement Proposal
## Observed Phenomenon
`MerkleBIT` struct has all those marker fields with `*const` which prevents it from auto-deriving `Sync+Send` when all other types (DatabaseType, BranchType, LeafType, DataType, NodeType, HasherType, ValueType) are `Sync+Send`.
Some of the internal functions take `*mut ...` arguments and have unsafe blocks in them.

## Proposed Process Improvement
I haven't looked into detail what's happening inside the unsafe blocks, but a lot of it didn't look fundamentally unachievable by using `&T` / `&mut T`.

Ideally, the crate would only use safe code or document the reason for not doing so.

**Comment by ChosunOne** [comment]
They aren't derived because the database itself must implement `Sync + Send` for the structure to be thread safe.  I've got two versions of the structure, one for parallel use, and another for single thread use, which can be toggled via the `parallel` feature in nightly.  

As for the unsafe sections, they are there because otherwise you would need to wrap the internal items in `Rc<RefCell<T>>` which has serious performance penalties. (You can't use `&t` or `&mut T` because of the loop involved, it would create multiple mutable borrows to the same item).  The unsafe sections are completely private, so I'm not worried about problems from specific inputs, as the unit tests and benchmarks are quite extensive.

**Comment by tomtau** [comment]
That parallel constraint is just for internal use and it only constrains the DatabaseType, NodeType, and ValueType to be `Sync+Send`... the other ones aren't.
And even with these constraints, `MerkleBIT` wouldn't be `Sync+Send`, because raw pointers (`*const`) are neither Send nor Sync.

Just to clarify, this is about the consequences for client code where one may want to share `MerkleBIT` among multiple threads (provided all its components, including the DB Type, are `Sync+Send`).

As for mutations in loops in unsafe sections, I'd need to look in detail to understand what they are trying to do -- from a quick look, they are computing some future state (by mutating the MerkleBIT-wide scope fields directly) that should be persisted later... from that, my hunch is that it could be rewritten not to require multiple mutable borrows and not to wrap everything in `Rc<RefCell<T>>` or things from `std::sync::*` -- one could have some local state to mutate and only one mutable borrow on the MerkleBIT parts at the end -- there may be some performance overhead (and the criterion benchmarks would uncover how significant), but if it's not too bad, I'll live happier with that and compiler-enforced safety guarantees than with hyper-optimized unit tested code :)

**Comment by tomtau** [comment]
@ChosunOne Anyway, there are two points raised in this issue:

1. Inability to use `MerkleBIT` in a multi-threaded client code. It currently derives `!Sync + !Send`, so one would need to work around that, e.g. by some unsafe type coercion black magic or by creating some single-threaded wrapper with a synchronous req-rep communication channel.
This should be easily fixable (`cargo test` passes here):
https://github.com/tomtau/merkle_bit/commit/b8087765ef0a0995ef8694d1833b53cfb1cc9202
Unless I missed something in the unsafe bits, it should be OK, as `&self` and `&mut self` should enforce a safe (multiple readers or a single writer) client use of `MerkleBIT`.
But not sure whether this change is acceptable -- there may be a reason not to have `Sync+Send` or only have `Send` on some of the traits... but this could potentially be feature-guarded.

What do you think?

2. Unsafe sections -- it may be trickier to get rid of them, but having a safe Rust-only implementation is IMHO desirable. If the performance impact is too big, this tradeoff could be documented and both versions would be available via feature flags.

**Comment by ChosunOne** [comment]
Thanks for distilling down your points from before.  I took a look at the code and was able to make the changes you suggested, removing unsafe blocks from the code which may have also boosted performance (Hooray!).  I've also removed the `*const` in front of the types to enable auto deriving the `Send + Sync` traits.  These updates are available in the latest nightly version and will be released once I finish the inclusion/exclusion proof additions.   

Currently the update is only applied to the non-parallel version of the tree, but you ought to be able to share the singled threaded version of the tree safely across threads (provided you implement `Safe + Sync` for the underlying data types).
